File: native/Sources/Isometry/Bridge/Optimization/QueryPaginator.swift
/**
 * QueryPaginator Actor for WebView Bridge Optimization
 *
 * Implements async query execution with cursor-based pagination using SQL LIMIT/OFFSET
 * with stable ordering, and integration with BinarySerializer for efficient transport.
 * Uses proper Swift concurrency patterns with actor isolation for thread safety.
 */

import Foundation
import GRDB
import OSLog

/// Represents a pagination cursor for stable pagination
public struct PaginationCursor: Sendable, Codable {
    public let value: String
    public let direction: PaginationDirection
    public let timestamp: TimeInterval

    public init(value: String, direction: PaginationDirection = .forward, timestamp: TimeInterval = Date().timeIntervalSince1970) {
        self.value = value
        self.direction = direction
        self.timestamp = timestamp
    }
}

/// Direction for pagination
public enum PaginationDirection: String, Sendable, Codable {
    case forward
    case backward
}

/// Configuration for paginated queries
public struct PaginatedQuery: Sendable {
    public let sql: String
    public let parameters: [DatabaseValue]
    public let limit: Int
    public let cursor: PaginationCursor?
    public let orderBy: String

    public init(sql: String, parameters: [DatabaseValue] = [], limit: Int = 50, cursor: PaginationCursor? = nil, orderBy: String = "id ASC") {
        self.sql = sql
        self.parameters = parameters
        self.limit = min(limit, 50) // Enforce 50 record maximum
        self.cursor = cursor
        self.orderBy = orderBy
    }
}

/// Result of a paginated query
public struct PaginatedResult<T: Sendable>: Sendable {
    public let data: [T]
    public let nextCursor: PaginationCursor?
    public let previousCursor: PaginationCursor?
    public let hasMore: Bool
    public let hasPrevious: Bool
    public let pageInfo: PageInfo

    public init(data: [T], nextCursor: PaginationCursor?, previousCursor: PaginationCursor?, hasMore: Bool, hasPrevious: Bool, pageInfo: PageInfo) {
        self.data = data
        self.nextCursor = nextCursor
        self.previousCursor = previousCursor
        self.hasMore = hasMore
        self.hasPrevious = hasPrevious
        self.pageInfo = pageInfo
    }
}

/// Page information metadata
public struct PageInfo: Sendable {
    public let currentPage: Int
    public let pageSize: Int
    public let isFirstPage: Bool
    public let isLastPage: Bool

    public init(currentPage: Int, pageSize: Int, isFirstPage: Bool, isLastPage: Bool) {
        self.currentPage = currentPage
        self.pageSize = pageSize
        self.isFirstPage = isFirstPage
        self.isLastPage = isLastPage
    }
}

/// Configuration options for QueryPaginator
public struct PaginatorOptions: Sendable {
    public let defaultPageSize: Int
    public let maxPageSize: Int
    public let enableEstimation: Bool
    public let cursorField: String
    public let cacheTimeout: TimeInterval

    public init(
        defaultPageSize: Int = 50,
        maxPageSize: Int = 50,
        enableEstimation: Bool = false,
        cursorField: String = "id",
        cacheTimeout: TimeInterval = 300 // 5 minutes
    ) {
        self.defaultPageSize = min(defaultPageSize, 50)
        self.maxPageSize = min(maxPageSize, 50)
        self.enableEstimation = enableEstimation
        self.cursorField = cursorField
        self.cacheTimeout = cacheTimeout
    }
}

/// Performance metrics for pagination operations
public struct PaginatorMetrics: Sendable {
    public let totalQueries: Int
    public let totalRecords: Int
    public let averagePageSize: Double
    public let averageQueryTime: TimeInterval
    public let cacheHits: Int
    public let cursorErrors: Int

    public init(totalQueries: Int, totalRecords: Int, averagePageSize: Double, averageQueryTime: TimeInterval, cacheHits: Int, cursorErrors: Int) {
        self.totalQueries = totalQueries
        self.totalRecords = totalRecords
        self.averagePageSize = averagePageSize
        self.averageQueryTime = averageQueryTime
        self.cacheHits = cacheHits
        self.cursorErrors = cursorErrors
    }
}

/// QueryPaginator actor for handling large dataset pagination
public actor QueryPaginator {
    // MARK: - Properties

    private let database: DatabaseQueue
    private let options: PaginatorOptions
    private let logger = Logger(subsystem: "IsometryBridge", category: "QueryPaginator")

    // Metrics tracking
    private var totalQueries = 0
    private var totalRecords = 0
    private var queryTimes: [TimeInterval] = []
    private var pageSizes: [Int] = []
    private var cacheHits = 0
    private var cursorErrors = 0

    // Cursor validation cache
    private var cursorCache: [String: (timestamp: TimeInterval, valid: Bool)] = [:]

    // MARK: - Initialization

    public init(database: DatabaseQueue, options: PaginatorOptions = PaginatorOptions()) {
        self.database = database
        self.options = options
    }

    // MARK: - Public Methods

    /// Execute a paginated query with cursor-based pagination
    public func executePaginatedQuery<T: FetchableRecord & Sendable>(
        query: PaginatedQuery,
        recordType: T.Type
    ) async throws -> PaginatedResult<T> {
        let startTime = CACurrentMediaTime()

        defer {
            let queryTime = CACurrentMediaTime() - startTime
            updateMetrics(pageSize: query.limit, queryTime: queryTime)
        }

        do {
            // Build paginated SQL
            let paginatedSQL = buildPaginatedSQL(query: query)

            // Execute query with one extra record to check for more pages
            let results = try await database.read { db in
                try recordType.fetchAll(db, sql: paginatedSQL.sql, arguments: StatementArguments(paginatedSQL.parameters))
            }

            // Process results and generate pagination info
            let paginationResult = try processPaginatedResults(
                results: results,
                originalQuery: query,
                queryTime: CACurrentMediaTime() - startTime
            )

            updateTotalQueries()

            return paginationResult

        } catch {
            incrementCursorErrors()
            throw QueryPaginatorError.paginationFailed(underlying: error)
        }
    }

    /// Create a paginated query from standard parameters
    public func createPaginatedQuery(
        sql: String,
        parameters: [DatabaseValue] = [],
        limit: Int? = nil,
        cursor: PaginationCursor? = nil,
        orderBy: String? = nil
    ) -> PaginatedQuery {
        let actualLimit = min(limit ?? options.defaultPageSize, options.maxPageSize)
        let actualOrderBy = orderBy ?? "\(options.cursorField) ASC"

        return PaginatedQuery(
            sql: sql,
            parameters: parameters,
            limit: actualLimit,
            cursor: cursor,
            orderBy: actualOrderBy
        )
    }

    /// Generate a stable cursor from a record
    public func generateCursor<T: FetchableRecord>(
        from record: T,
        direction: PaginationDirection = .forward
    ) throws -> PaginationCursor {
        // Use reflection to get cursor field value
        let mirror = Mirror(reflecting: record)

        for child in mirror.children {
            if let label = child.label, label == options.cursorField {
                if let value = child.value as? CustomStringConvertible {
                    return PaginationCursor(
                        value: value.description,
                        direction: direction,
                        timestamp: Date().timeIntervalSince1970
                    )
                }
            }
        }

        throw QueryPaginatorError.cursorFieldNotFound(field: options.cursorField)
    }

    /// Validate cursor stability and freshness
    public func validateCursor(_ cursor: PaginationCursor) async throws -> Bool {
        let cacheKey = "\(cursor.value)_\(cursor.direction.rawValue)_\(cursor.timestamp)"

        // Check cache first
        if let cached = cursorCache[cacheKey],
           Date().timeIntervalSince1970 - cached.timestamp < options.cacheTimeout {
            if cached.valid {
                incrementCacheHits()
            }
            return cached.valid
        }

        do {
            // Verify cursor exists in database
            let exists = try await database.read { [self] db in
                try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM (SELECT \(self.options.cursorField) FROM nodes WHERE \(self.options.cursorField) = ?)", arguments: [cursor.value]) ?? 0 > 0
            }

            // Cache the result
            cursorCache[cacheKey] = (timestamp: Date().timeIntervalSince1970, valid: exists)

            // Clean old cache entries
            cleanCursorCache()

            return exists
        } catch {
            incrementCursorErrors()
            return false
        }
    }

    /// Get next page using cursor
    public func getNextPage<T: FetchableRecord & Sendable>(
        from currentResult: PaginatedResult<T>,
        originalQuery: PaginatedQuery,
        recordType: T.Type
    ) async throws -> PaginatedResult<T>? {
        guard currentResult.hasMore, let nextCursor = currentResult.nextCursor else {
            return nil
        }

        let nextQuery = PaginatedQuery(
            sql: originalQuery.sql,
            parameters: originalQuery.parameters,
            limit: originalQuery.limit,
            cursor: nextCursor,
            orderBy: originalQuery.orderBy
        )

        return try await executePaginatedQuery(query: nextQuery, recordType: recordType)
    }

    /// Get previous page using cursor
    public func getPreviousPage<T: FetchableRecord & Sendable>(
        from currentResult: PaginatedResult<T>,
        originalQuery: PaginatedQuery,
        recordType: T.Type
    ) async throws -> PaginatedResult<T>? {
        guard currentResult.hasPrevious, let previousCursor = currentResult.previousCursor else {
            return nil
        }

        let prevQuery = PaginatedQuery(
            sql: originalQuery.sql,
            parameters: originalQuery.parameters,
            limit: originalQuery.limit,
            cursor: previousCursor,
            orderBy: originalQuery.orderBy
        )

        return try await executePaginatedQuery(query: prevQuery, recordType: recordType)
    }

    /// Get current paginator metrics
    public func getMetrics() -> PaginatorMetrics {
        let averagePageSize = pageSizes.isEmpty ? 0.0 : Double(pageSizes.reduce(0, +)) / Double(pageSizes.count)
        let averageQueryTime = queryTimes.isEmpty ? 0.0 : queryTimes.reduce(0, +) / Double(queryTimes.count)

        return PaginatorMetrics(
            totalQueries: totalQueries,
            totalRecords: totalRecords,
            averagePageSize: averagePageSize,
            averageQueryTime: averageQueryTime,
            cacheHits: cacheHits,
            cursorErrors: cursorErrors
        )
    }

    /// Reset metrics (useful for testing)
    public func resetMetrics() {
        totalQueries = 0
        totalRecords = 0
        queryTimes = []
        pageSizes = []
        cacheHits = 0
        cursorErrors = 0
        cursorCache = [:]
    }

    /// Clear cursor cache
    public func clearCache() {
        cursorCache = [:]
    }

    // MARK: - Private Methods

    private func buildPaginatedSQL(query: PaginatedQuery) -> (sql: String, parameters: [DatabaseValue]) {
        var sql = query.sql
        var parameters = query.parameters

        // Add cursor condition if provided
        if let cursor = query.cursor {
            let comparisonOperator = cursor.direction == .forward ? ">" : "<"
            let cursorCondition = "\(options.cursorField) \(comparisonOperator) ?"

            if sql.lowercased().contains("where") {
                sql += " AND \(cursorCondition)"
            } else {
                sql += " WHERE \(cursorCondition)"
            }

            parameters.append(DatabaseValue.text(cursor.value))
        }

        // Add ordering (ensure consistent ordering for stable pagination)
        if !sql.lowercased().contains("order by") {
            sql += " ORDER BY \(query.orderBy)"
        }

        // Add limit (request one extra record to check for more pages)
        sql += " LIMIT \(query.limit + 1)"

        return (sql: sql, parameters: parameters)
    }

    private func processPaginatedResults<T: FetchableRecord & Sendable>(
        results: [T],
        originalQuery: PaginatedQuery,
        queryTime: TimeInterval
    ) throws -> PaginatedResult<T> {
        let hasMore = results.count > originalQuery.limit
        let data = hasMore ? Array(results.dropLast()) : results

        var nextCursor: PaginationCursor?
        var previousCursor: PaginationCursor?

        // Generate next cursor if there are more records
        if hasMore, let lastRecord = data.last {
            nextCursor = try generateCursor(from: lastRecord, direction: .forward)
        }

        // Generate previous cursor if this isn't the first page
        if originalQuery.cursor != nil, let firstRecord = data.first {
            previousCursor = try generateCursor(from: firstRecord, direction: .backward)
        }

        // Estimate current page number (rough approximation)
        let currentPage = originalQuery.cursor != nil ?
            max(Int(queryTime / (queryTimes.isEmpty ? 0.1 : queryTimes.reduce(0, +) / Double(queryTimes.count))) + 1, 1) : 1

        let pageInfo = PageInfo(
            currentPage: currentPage,
            pageSize: data.count,
            isFirstPage: originalQuery.cursor == nil,
            isLastPage: !hasMore
        )

        return PaginatedResult(
            data: data,
            nextCursor: nextCursor,
            previousCursor: previousCursor,
            hasMore: hasMore,
            hasPrevious: originalQuery.cursor != nil,
            pageInfo: pageInfo
        )
    }

    private func cleanCursorCache() {
        let now = Date().timeIntervalSince1970
        cursorCache = cursorCache.filter { _, entry in
            now - entry.timestamp < options.cacheTimeout
        }
    }

    private func updateMetrics(pageSize: Int, queryTime: TimeInterval) {
        totalRecords += pageSize

        // Keep last 100 measurements
        queryTimes.append(queryTime)
        pageSizes.append(pageSize)

        if queryTimes.count > 100 {
            queryTimes.removeFirst()
        }
        if pageSizes.count > 100 {
            pageSizes.removeFirst()
        }
    }

    private func updateTotalQueries() {
        totalQueries += 1
    }

    private func incrementCacheHits() {
        cacheHits += 1
    }

    private func incrementCursorErrors() {
        cursorErrors += 1
    }
}

// MARK: - Error Types

public enum QueryPaginatorError: Error {
    case paginationFailed(underlying: Error)
    case cursorFieldNotFound(field: String)
    case invalidCursor(reason: String)
}

// MARK: - Extensions

extension QueryPaginatorError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .paginationFailed(let error):
            return "Pagination query failed: \(error.localizedDescription)"
        case .cursorFieldNotFound(let field):
            return "Cursor field '\(field)' not found in record"
        case .invalidCursor(let reason):
            return "Invalid cursor: \(reason)"
        }
    }
}

File: src/utils/bridge-optimization/query-paginator.ts
/**
 * Query Paginator for WebView Bridge Optimization
 *
 * Implements cursor-based pagination for large datasets with 50-record page limits,
 * stable cursor generation using record IDs, and integration with MessageBatcher and
 * BinarySerializer for optimized transport.
 */

export interface PaginationCursor {
  value: string;
  direction: 'forward' | 'backward';
  timestamp: number;
}

export interface PaginatedQuery {
  sql: string;
  params: unknown[];
  limit: number;
  cursor?: PaginationCursor;
  orderBy: string;
}

export interface PaginatedResult<T = unknown> {
  data: T[];
  nextCursor?: PaginationCursor;
  previousCursor?: PaginationCursor;
  hasMore: boolean;
  hasPrevious: boolean;
  totalEstimate?: number;
  pageInfo: {
    currentPage: number;
    pageSize: number;
    isFirstPage: boolean;
    isLastPage: boolean;
  };
}

export interface PaginatorOptions {
  defaultPageSize?: number;
  maxPageSize?: number;
  enableEstimation?: boolean;
  cursorField?: string;
}

export interface PaginatorMetrics {
  totalQueries: number;
  totalRecords: number;
  averagePageSize: number;
  averageQueryTime: number;
  cacheHits: number;
  cursorErrors: number;
}

/**
 * QueryPaginator manages cursor-based pagination for large database queries
 */
export class QueryPaginator {
  private readonly defaultPageSize: number;
  private readonly maxPageSize: number;
  // Legacy option - preserved for compatibility
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private readonly _enableEstimation: boolean;
  private readonly cursorField: string;

  // Metrics tracking
  private metrics: PaginatorMetrics = {
    totalQueries: 0,
    totalRecords: 0,
    averagePageSize: 0,
    averageQueryTime: 0,
    cacheHits: 0,
    cursorErrors: 0
  };

  // Performance tracking arrays
  private queryTimes: number[] = [];
  private pageSizes: number[] = [];

  // Cursor validation cache
  private cursorCache = new Map<string, { timestamp: number; valid: boolean }>();
  private readonly cursorCacheTimeout = 5 * 60 * 1000; // 5 minutes

  // Query execution function - injected for testability and bridge integration
  private executeQuery: (sql: string, params: unknown[]) => Promise<unknown[]>;

  constructor(
    executeQuery: (sql: string, params: unknown[]) => Promise<unknown[]>,
    options: PaginatorOptions = {}
  ) {
    this.executeQuery = executeQuery;
    this.defaultPageSize = Math.min(options.defaultPageSize ?? 50, 50); // Max 50 records per message
    this.maxPageSize = Math.min(options.maxPageSize ?? 50, 50); // Enforce 50 record limit
    this._enableEstimation = options.enableEstimation ?? false;
    this.cursorField = options.cursorField ?? 'id';

    // Explicitly mark preserved fields
    void this._enableEstimation;
  }

  /**
   * Execute a paginated query with cursor-based pagination
   */
  async executePaginatedQuery<T = unknown>(
    query: PaginatedQuery,
    requestedPageSize?: number
  ): Promise<PaginatedResult<T>> {
    const startTime = performance.now();

    try {
      // Validate and limit page size
      const pageSize = Math.min(requestedPageSize ?? query.limit ?? this.defaultPageSize, this.maxPageSize);

      // Build paginated SQL query
      const paginatedSQL = this.buildPaginatedSQL(query, pageSize);

      // Execute query with one extra record to check for more pages
      const results = await this.executeQuery(paginatedSQL.sql, paginatedSQL.params) as T[];

      const queryTime = performance.now() - startTime;

      // Process results and generate pagination info
      const paginationResult = this.processPaginatedResults(
        results,
        query,
        pageSize,
        queryTime
      );

      // Update metrics
      this.updateMetrics(pageSize, queryTime);

      return paginationResult;

    } catch (error) {
      this.metrics.cursorErrors++;
      throw new Error(`Pagination query failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Create a paginated query from a standard query
   */
  createPaginatedQuery(
    sql: string,
    params: unknown[] = [],
    options: {
      limit?: number;
      cursor?: PaginationCursor;
      orderBy?: string;
    } = {}
  ): PaginatedQuery {
    const limit = Math.min(options.limit ?? this.defaultPageSize, this.maxPageSize);
    const orderBy = options.orderBy ?? `${this.cursorField} ASC`;

    return {
      sql,
      params,
      limit,
      cursor: options.cursor,
      orderBy
    };
  }

  /**
   * Generate a stable cursor from a record
   */
  generateCursor(
    record: Record<string, unknown>,
    direction: 'forward' | 'backward' = 'forward'
  ): PaginationCursor {
    const cursorValue = record[this.cursorField];

    if (cursorValue === null || cursorValue === undefined) {
      throw new Error(`Cursor field '${this.cursorField}' not found in record`);
    }

    return {
      value: String(cursorValue),
      direction,
      timestamp: Date.now()
    };
  }

  /**
   * Validate cursor stability and freshness
   */
  async validateCursor(cursor: PaginationCursor): Promise<boolean> {
    const cacheKey = `${cursor.value}_${cursor.direction}_${cursor.timestamp}`;

    // Check cache first
    const cached = this.cursorCache.get(cacheKey);
    if (cached && (Date.now() - cached.timestamp) < this.cursorCacheTimeout) {
      if (cached.valid) this.metrics.cacheHits++;
      return cached.valid;
    }

    try {
      // Verify cursor exists in database
      const checkSQL = `SELECT 1 FROM (${this.getBaseSQLFromQuery()}) WHERE ${this.cursorField} = ?`;
      const results = await this.executeQuery(checkSQL, [cursor.value]);
      const isValid = results.length > 0;

      // Cache the result
      this.cursorCache.set(cacheKey, { timestamp: Date.now(), valid: isValid });

      // Clean old cache entries
      this.cleanCursorCache();

      return isValid;
    } catch (error) {
      this.metrics.cursorErrors++;
      return false;
    }
  }

  /**
   * Get next page using cursor
   */
  async getNextPage<T = unknown>(
    currentResult: PaginatedResult<T>,
    originalQuery: PaginatedQuery
  ): Promise<PaginatedResult<T> | null> {
    if (!currentResult.hasMore || !currentResult.nextCursor) {
      return null;
    }

    const nextQuery: PaginatedQuery = {
      ...originalQuery,
      cursor: currentResult.nextCursor
    };

    return this.executePaginatedQuery(nextQuery);
  }

  /**
   * Get previous page using cursor
   */
  async getPreviousPage<T = unknown>(
    currentResult: PaginatedResult<T>,
    originalQuery: PaginatedQuery
  ): Promise<PaginatedResult<T> | null> {
    if (!currentResult.hasPrevious || !currentResult.previousCursor) {
      return null;
    }

    const prevQuery: PaginatedQuery = {
      ...originalQuery,
      cursor: currentResult.previousCursor
    };

    return this.executePaginatedQuery(prevQuery);
  }

  /**
   * Build SQL query with cursor-based pagination
   */
  private buildPaginatedSQL(query: PaginatedQuery, pageSize: number): {
    sql: string;
    params: unknown[];
  } {
    const { sql, params, cursor, orderBy } = query;

    let paginatedSQL = sql;
    const paginatedParams = [...params];

    // Add cursor condition if provided
    if (cursor) {
      const operator = cursor.direction === 'forward' ? '>' : '<';
      const cursorCondition = this.buildCursorCondition(sql, operator, cursor.value);
      paginatedSQL = this.addWhereCondition(sql, cursorCondition);
      paginatedParams.push(cursor.value);
    }

    // Add ordering (ensure consistent ordering for stable pagination)
    if (!sql.toLowerCase().includes('order by')) {
      paginatedSQL += ` ORDER BY ${orderBy}`;
    }

    // Add limit (request one extra record to check for more pages)
    paginatedSQL += ` LIMIT ${pageSize + 1}`;

    return {
      sql: paginatedSQL,
      params: paginatedParams
    };
  }

  /**
   * Process paginated query results
   */
  private processPaginatedResults<T = unknown>(
    results: T[],
    originalQuery: PaginatedQuery,
    pageSize: number,
    queryTime: number
  ): PaginatedResult<T> {
    const hasMore = results.length > pageSize;
    const data = hasMore ? results.slice(0, pageSize) : results;

    let nextCursor: PaginationCursor | undefined;
    let previousCursor: PaginationCursor | undefined;

    // Generate next cursor if there are more records
    if (hasMore && data.length > 0) {
      const lastRecord = data[data.length - 1] as Record<string, unknown>;
      nextCursor = this.generateCursor(lastRecord, 'forward');
    }

    // Generate previous cursor if this isn't the first page
    if (originalQuery.cursor && data.length > 0) {
      const firstRecord = data[0] as Record<string, unknown>;
      previousCursor = this.generateCursor(firstRecord, 'backward');
    }

    // Estimate current page number (rough approximation)
    const currentPage = originalQuery.cursor ?
      Math.floor(queryTime / this.metrics.averageQueryTime) || 1 : 1;

    return {
      data,
      nextCursor,
      previousCursor,
      hasMore,
      hasPrevious: !!originalQuery.cursor,
      pageInfo: {
        currentPage,
        pageSize: data.length,
        isFirstPage: !originalQuery.cursor,
        isLastPage: !hasMore
      }
    };
  }

  /**
   * Build cursor condition for WHERE clause
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private buildCursorCondition(_sql: string, operator: string, _cursorValue: string): string {
    return `${this.cursorField} ${operator} ?`;
  }

  /**
   * Add WHERE condition to existing SQL
   */
  private addWhereCondition(sql: string, condition: string): string {
    const lowerSQL = sql.toLowerCase();

    if (lowerSQL.includes('where')) {
      return sql + ` AND ${condition}`;
    } else {
      return sql + ` WHERE ${condition}`;
    }
  }

  /**
   * Extract base SQL for cursor validation (simplified)
   */
  private getBaseSQLFromQuery(): string {
    // Simplified implementation - in practice, would parse the query more carefully
    return "SELECT id FROM nodes";
  }

  /**
   * Update performance metrics
   */
  private updateMetrics(pageSize: number, queryTime: number): void {
    this.metrics.totalQueries++;
    this.metrics.totalRecords += pageSize;

    // Track performance arrays (keep last 100 measurements)
    this.queryTimes.push(queryTime);
    this.pageSizes.push(pageSize);

    if (this.queryTimes.length > 100) {
      this.queryTimes.shift();
    }
    if (this.pageSizes.length > 100) {
      this.pageSizes.shift();
    }

    // Update averages
    this.metrics.averageQueryTime =
      this.queryTimes.reduce((a, b) => a + b, 0) / this.queryTimes.length;
    this.metrics.averagePageSize =
      this.pageSizes.reduce((a, b) => a + b, 0) / this.pageSizes.length;
  }

  /**
   * Clean expired cursor cache entries
   */
  private cleanCursorCache(): void {
    const now = Date.now();
    const keysToDelete: string[] = [];

    this.cursorCache.forEach((entry, key) => {
      if (now - entry.timestamp > this.cursorCacheTimeout) {
        keysToDelete.push(key);
      }
    });

    keysToDelete.forEach(key => {
      this.cursorCache.delete(key);
    });
  }

  /**
   * Get current paginator metrics
   */
  getMetrics(): PaginatorMetrics {
    return { ...this.metrics };
  }

  /**
   * Reset metrics (useful for testing)
   */
  resetMetrics(): void {
    this.metrics = {
      totalQueries: 0,
      totalRecords: 0,
      averagePageSize: 0,
      averageQueryTime: 0,
      cacheHits: 0,
      cursorErrors: 0
    };
    this.queryTimes = [];
    this.pageSizes = [];
  }

  /**
   * Get pagination performance report
   */
  getPerformanceReport(): {
    efficiency: {
      averageQueryTime: number;
      averagePageSize: number;
      cacheHitRate: number;
      errorRate: number;
    };
    throughput: {
      queriesPerSecond: number;
      recordsPerSecond: number;
    };
    health: {
      isPerformingWell: boolean;
      recommendations: string[];
    };
  } {
    const cacheHitRate = this.metrics.totalQueries > 0 ?
      (this.metrics.cacheHits / this.metrics.totalQueries) : 0;
    const errorRate = this.metrics.totalQueries > 0 ?
      (this.metrics.cursorErrors / this.metrics.totalQueries) : 0;

    const recommendations: string[] = [];

    if (this.metrics.averageQueryTime > 100) {
      recommendations.push('High query times detected - consider adding database indexes');
    }

    if (cacheHitRate < 0.3) {
      recommendations.push('Low cursor cache hit rate - consider increasing cache timeout');
    }

    if (errorRate > 0.05) {
      recommendations.push('High cursor error rate - check cursor stability and database consistency');
    }

    if (this.metrics.averagePageSize < this.defaultPageSize * 0.5) {
      recommendations.push('Low average page size - consider adjusting query filters or pagination strategy');
    }

    const isPerformingWell = this.metrics.averageQueryTime < 50 &&
                           cacheHitRate > 0.5 &&
                           errorRate < 0.01;

    return {
      efficiency: {
        averageQueryTime: this.metrics.averageQueryTime,
        averagePageSize: this.metrics.averagePageSize,
        cacheHitRate,
        errorRate
      },
      throughput: {
        queriesPerSecond: this.metrics.averageQueryTime > 0 ?
          1000 / this.metrics.averageQueryTime : 0,
        recordsPerSecond: this.metrics.averageQueryTime > 0 ?
          (this.metrics.averagePageSize * 1000) / this.metrics.averageQueryTime : 0
      },
      health: {
        isPerformingWell,
        recommendations
      }
    };
  }

  /**
   * Clear cursor cache
   */
  clearCache(): void {
    this.cursorCache.clear();
  }
}